consensus/core/src/core.rs

```diff
@@ -111,47 +154,196 @@ impl Core {
+            // TODO: this will be refactored once the signing path/approach has been introduced. Adding as is for now
+            // to keep things rolling in the implementation.
+            let block = Block::V1(BlockV1::new(
+                clock_round,
+                self.context.own_index,
+                now,
+                ancestors,
+                payload,
+                self.context.committee.epoch(),
+            ));
+            let signed_block = SignedBlock::new(block);
+            let verified_block = VerifiedBlock::new_verified_unserialized(signed_block)
+                .expect("Fatal error, creating a verified block failed");
+
+            //4. Add to the threshold clock
+            self.threshold_clock.add_block(verified_block.reference());
+
+            // TODO: use the DagState instead to accept the block directly. Will address on follow up PR when I'll inject
+            // the DagState. Now that's necessary to ensure that blocks aren't double processed.
+            self.block_manager.add_blocks(vec![verified_block.clone()]);
+
+            self.last_own_block = verified_block.clone();
+
+            tracing::debug!("New block created {}", verified_block);
+
+            //5. emit an event that a new block is ready
+            self.signals.new_block_ready(verified_block.reference());
+
+            return Some(verified_block);
         }
 
         None
     }
 
-    fn ready_new_block(&self) -> bool {
-        // TODO: check that we are ready to produce a new block. This will mainly check that the leader of the previous
-        // quorum exists.
-        true
+    /// Retrieves the next ancestors to propose to form a block at `clock_round` round. Also the `block_timestamp` is provided
+    /// to sanity check that everything that goes into the proposal is ensured to have a timestamp < block_timestamp
+    fn ancestors_to_propose(
+        &mut self,
+        clock_round: Round,
+        block_timestamp: BlockTimestampMs,
+    ) -> Vec<BlockRef> {
+        // Now take all the ancestors up to the clock_round (excluded) and then filter only the ones with acceptable timestamp.
+        let first_include_index = self
+            .pending_ancestors
+            .iter()
+            .position(|(block_ref, timestamp)| {
+                // We assume that our system's clock can't go backwards when we perform the check here (ex due to ntp corrections)
+                assert!(*timestamp <= block_timestamp, "Violation, ancestor block timestamp {timestamp} greater than our timestamp {block_timestamp}");
+
+                block_ref.round >= clock_round
+            })
+            .unwrap_or(self.pending_ancestors.len());
+
+        let mut taken = self.pending_ancestors.split_off(first_include_index);
+        mem::swap(&mut taken, &mut self.pending_ancestors);
+
+        // Compress the references in the block. We don't want to include an ancestors that already referenced by other blocks
```

Comment:
1. There should be a short comment on why no invariant is violated, e.g. why parents at round r-1 should be unaffected.
2. This logic seems to produce an interesting dynamic: if validator A stopped producing new blocks, peers will alternative between including and excluding the ref to A's last block. An alternative can be to remove refs already exist in the last own block, which I think is easier to reason about.

Reply:
When reading about ancestor compression, I would wonder if references to last own block or 2f+1 parents can be compressed as well. It is a good point that ancestors removed from pending will never get put back again.